README.md
# HAI Server [![GitHub issues](https://img.shields.io/github/issues/sumn2u/hai-server.svg)](https://github.com/sumn2u/hai-server/issues) [![GitHub forks](https://img.shields.io/github/forks/sumn2u/hai-server.svg)](https://github.com/sumn2u/hai-server/network) [![GitHub stars](https://img.shields.io/github/stars/sumn2u/hai-server.svg)](https://github.com/sumn2u/hai-server/stargazers) [![GitHub license](https://img.shields.io/github/license/sumn2u/hai-server.svg)](https://github.com/sumn2u/hai-server/blob/master/LICENSE) [![Twitter](https://img.shields.io/twitter/url/https/github.com/sumn2u/hai-server.svg?style=social)](https://twitter.com/intent/tweet?text=Wow:&url=https%3A%2F%2Fgithub.com%2Fsumn2u%2Fhai-server)
[![GitHub tag](https://img.shields.io/github/tag/sumn2u/hai-server.svg)](https://GitHub.com/sumn2u/hai-server/tags/)
[![GitHub release](https://img.shields.io/github/release/sumn2u/hai-server.svg)](https://GitHub.com/sumn2u/hai-server/releases/)
[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/sumn2u/hai-server/graphs/commit-activity)
<a href="https://npmjs.com/package/money-cli"><img src="https://img.shields.io/npm/dt/hai-server.svg" alt="npm Downloads"></a> 


<!-- tocstop -->
Inspired and Forked from [JSON Server](https://github.com/typicode/json-server)

Get a full fake REST API with authorization.

## Getting started

Install HAI Server 

```
npm install -g hai-server
```

Create a `db.json` file with some data

```json
{
  "posts": [
    { "id": 1, "title": "hai-server", "author": "sumn2u" }
  ],
  "comments": [
    { "id": 1, "body": "some comment", "postId": 1 }
  ],
  "profile": { "name": "sumn2u" }
}
```

Start HAI Server

```bash
hai-server --watch db.json
```

Now if you go to [http://localhost:3000/posts/1](http://localhost:3000/posts/1), you'll get

```json
{ "id": 1, "title": "hai-server", "author": "sumn2u" }
```

Also when doing requests, it's good to know that:

- If you make POST, PUT, PATCH or DELETE requests, changes will be automatically and safely saved to `db.json` using [lowdb](https://github.com/sumn2u/lowdb).
- Your request body JSON should be object enclosed, just like the GET output. (for example `{"name": "Foobar"}`)
- Id values are not mutable. Any `id` value in the body of your PUT or PATCH request will be ignored. Only a value set in a POST request will be respected, but only if not already taken.
- A POST, PUT or PATCH request should include a `Content-Type: application/json` header to use the JSON in the request body. Otherwise it will result in a 200 OK but without changes being made to the data.

## Routes

Based on the previous `db.json` file, here are all the default routes. You can also add [other routes](#add-custom-routes) using `--routes`.

### Plural routes

```
GET    /posts
GET    /posts/1
POST   /posts
PUT    /posts/1
PATCH  /posts/1
DELETE /posts/1
```

### Singular routes

```
GET    /profile
POST   /profile
PUT    /profile
PATCH  /profile
```

### Filter

Use `.` to access deep properties

```
GET /posts?title=hai-server&author=sumn2u
GET /posts?id=1&id=2
GET /comments?author.name=sumn2u
```

### Paginate

Use `_page` and optionally `_limit` to paginate returned data.

In the `Link` header you'll get `first`, `prev`, `next` and `last` links.


```
GET /posts?_page=7
GET /posts?_page=7&_limit=20
```

_10 items are returned by default_

### Sort

Add `_sort` and `_order` (ascending order by default)

```
GET /posts?_sort=views&_order=asc
GET /posts/1/comments?_sort=votes&_order=asc
```

For multiple fields, use the following format:

```
GET /posts?_sort=user,views&_order=desc,asc
```

### Slice

Add `_start` and `_end` or `_limit` (an `X-Total-Count` header is included in the response)

```
GET /posts?_start=20&_end=30
GET /posts/1/comments?_start=20&_end=30
GET /posts/1/comments?_start=20&_limit=10
```

_Works exactly as [Array.slice](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Array/slice) (i.e. `_start` is inclusive and `_end` exclusive)_

### Operators

Add `_gte` or `_lte` for getting a range

```
GET /posts?views_gte=10&views_lte=20
```

Add `_ne` to exclude a value

```
GET /posts?id_ne=1
```

Add `_like` to filter (RegExp supported)

```
GET /posts?title_like=server
```

### Full-text search

Add `q`

```
GET /posts?q=internet
```

### Relationships

To include children resources, add `_embed`

```
GET /posts?_embed=comments
GET /posts/1?_embed=comments
```

To include parent resource, add `_expand`

```
GET /comments?_expand=post
GET /comments/1?_expand=post
```

To get or create nested resources (by default one level, [add custom routes](#add-custom-routes) for more)

```
GET  /posts/1/comments
POST /posts/1/comments
```

### Database

```
GET /db
```

### Homepage

Returns default index file or serves `./public` directory

```
GET /
```

## Extras

### Static file server

You can use HAI Server to serve your HTML, JS and CSS, simply create a `./public` directory
or use `--static` to set a different static files directory.

```bash
mkdir public
echo 'hello world' > public/index.html
hai-server db.json
```

```bash
hai-server db.json --static ./some-other-dir
```

### Alternative port

You can start HAI Server on other ports with the `--port` flag:

```bash
$ hai-server --watch db.json --port 3004
```

### Access from anywhere

You can access your fake API from anywhere using CORS and JSONP.

### Remote schema

You can load remote schemas.

```bash
$ hai-server http://example.com/file.json
$ hai-server http://jsonplaceholder.sumn2u.com/db
```

### Generate random data

Using JS instead of a JSON file, you can create data programmatically.

```javascript
// index.js
module.exports = () => {
  const data = { users: [] }
  // Create 1000 users
  for (let i = 0; i < 1000; i++) {
    data.users.push({ id: i, name: `user${i}` })
  }
  return data
}
```

```bash
$ hai-server index.js
```

__Tip__ use modules like [Faker](https://github.com/Marak/faker.js), [Casual](https://github.com/boo1ean/casual), [Chance](https://github.com/victorquinn/chancejs) or [JSON Schema Faker](https://github.com/json-schema-faker/json-schema-faker).

### HTTPS

There are many ways to set up SSL in development. One simple way is to use [hotel](https://github.com/sumn2u/hotel).

### Add custom routes

Create a `routes.json` file. Pay attention to start every route with `/`.

```json
{
  "/api/*": "/$1",
  "/:resource/:id/show": "/:resource/:id",
  "/posts/:category": "/posts?category=:category",
  "/articles\\?id=:id": "/posts/:id"
}
```

Start HAI Server with `--routes` option.

```bash
hai-server db.json --routes routes.json
```

Now you can access resources using additional routes.

```sh
/api/posts # → /posts
/api/posts/1  # → /posts/1
/posts/1/show # → /posts/1
/posts/javascript # → /posts?category=javascript
/articles?id=1 # → /posts/1
```

### Add middlewares

You can add your middlewares from the CLI using `--middlewares` option:

```js
// hello.js
module.exports = (req, res, next) => {
  res.header('X-Hello', 'World')
  next()
}
```

```bash
hai-server db.json --middlewares ./hello.js
hai-server db.json --middlewares ./first.js ./second.js
```

### CLI usage

```
hai-server [options] <source>

Options:
  --config, -c       Path to config file           [default: "hai-server.json"]
  --port, -p         Set port                                    [default: 3000]
  --host, -H         Set host                             [default: "localhost"]
  --watch, -w        Watch file(s)                                     [boolean]
  --routes, -r       Path to routes file
  --middlewares, -m  Paths to middleware files                           [array]
  -- auth, -a        Authentication file                               
  --static, -s       Set static files directory
  --read-only, --ro  Allow only GET requests                           [boolean]
  --no-cors, --nc    Disable Cross-Origin Resource Sharing             [boolean]
  --no-gzip, --ng    Disable GZIP Content-Encoding                     [boolean]
  --snapshots, -S    Set snapshots directory                      [default: "."]
  --delay, -d        Add delay to responses (ms)
  --id, -i           Set database id property (e.g. _id)         [default: "id"]
  --foreignKeySuffix, --fks  Set foreign key suffix, (e.g. _id as in post_id)
                                                                 [default: "Id"]
  --quiet, -q        Suppress log messages from output                 [boolean]
  --help, -h         Show help                                         [boolean]
  --version, -v      Show version number                               [boolean]

Examples:
  hai-server db.json
  hai-server file.js
  hai-server http://example.com/db.json

https://github.com/sumn2u/hai-server
```

You can also set options in a `hai-server.json` configuration file.

```json
{
  "port": 3000
}
```

### Module

If you need to add authentication, validation, or __any behavior__, you can use the project as a module in combination with other Express middlewares.

#### Simple example

```sh
$ npm install hai-server --save-dev
```

```js
// server.js
const jsonServer = require('hai-server').haijsServer
const server = jsonServer.create()
const router = jsonServer.router('db.json')
const middlewares = jsonServer.defaults()

server.use(middlewares)
server.use(router)
server.listen(3000, () => {
  console.log('HAI Server is running')
})
```

```sh
$ node server.js
```

The path you provide to the `jsonServer.router` function  is relative to the directory from where you launch your node process. If you run the above code from another directory, it’s better to use an absolute path:

```js
const path = require('path')
const router = jsonServer.router(path.join(__dirname, 'db.json'))
```

For an in-memory database, simply pass an object to `jsonServer.router()`.

Please note also that `jsonServer.router()` can be used in existing Express projects.

#### Custom routes example

Let's say you want a route that echoes query parameters and another one that set a timestamp on every resource created.

```js
const jsonServer = require('hai-server').haijsServer
const server = jsonServer.create()
const router = jsonServer.router('db.json')
const middlewares = jsonServer.defaults()

// Set default middlewares (logger, static, cors and no-cache)
server.use(middlewares)

// Add custom routes before HAI Server router
server.get('/echo', (req, res) => {
  res.jsonp(req.query)
})

// To handle POST, PUT and PATCH you need to use a body-parser
// You can use the one used by HAI Server
server.use(jsonServer.bodyParser)
server.use((req, res, next) => {
  if (req.method === 'POST') {
    req.body.createdAt = Date.now()
  }
  // Continue to HAI Server router
  next()
})

// Use default router
server.use(router)
server.listen(3000, () => {
  console.log('HAI Server is running')
})
```  

#### Access control example
Create a file name `auth.json` similar like the one shown below. Provide your own `secretKey`, `expiresIn`, and collection of authenticated users. Here the `authenticatedURL` indicates the collection of URL's that needs authorization and `authenticationURL` is the URL to get accesstoken.

```json
{
    "secretKey":"123456789",
    "expiresIn": "1h",
    "users": [
        {
          "id": 1,
          "name": "bruno",
          "email": "bruno@email.com",
          "password": "bruno"
        },
        {
          "id": 2,
          "name": "techie",
          "email": "techie@email.com",
          "password": "techie"
        },
        {
          "id": 3,
          "name": "nilson",
          "email": "nilson@email.com",
          "password": "nilson"
        }
      ],
    "authenticationURL":"/auth/login",
    "authenticatedURL":["/users"]

}
```
```
hai-server --watch db.json --auth auth.json --port 3000
```
## How to login?

You can login by sending a POST request to

```
POST http://localhost:3000/auth/login
```
with the following data 

```
{
  "email": "nilson@email.com",
  "password":"nilson"
}
```

You should receive an access token with the following format 

```
{
   "access_token": "<ACCESS_TOKEN>"
}
```


You should send this authorization with any request to the protected endpoints

```
Authorization: Bearer <ACCESS_TOKEN>
```


#### Custom output example

To modify responses, overwrite `router.render` method:

```javascript
// In this example, returned resources will be wrapped in a body property
router.render = (req, res) => {
  res.jsonp({
    body: res.locals.data
  })
}
```

You can set your own status code for the response:


```javascript
// In this example we simulate a server side error response
router.render = (req, res) => {
  res.status(500).jsonp({
    error: "error message here"
  })
}
```

#### Rewriter example

To add rewrite rules, use `jsonServer.rewriter()`:

```javascript
// Add this before server.use(router)
server.use(jsonServer.rewriter({
  '/api/*': '/$1',
  '/blog/:resource/:id/show': '/:resource/:id'
}))
```

#### Mounting HAI Server on another endpoint example

Alternatively, you can also mount the router on `/api`.

```javascript
server.use('/api', router)
```

#### API

__`jsonServer.create()`__

Returns an Express server.

__`jsonServer.defaults([options])`__

Returns middlewares used by HAI Server.

* options
  * `static` path to static files
  * `logger` enable logger middleware (default: true)
  * `bodyParser` enable body-parser middleware (default: true)
  * `noCors` disable CORS (default: false)
  * `readOnly` accept only GET requests (default: false)

__`jsonServer.router([path|object])`__

Returns HAI Server router.

### Deployment


## Links


## License

MIT



CHANGELOG.md
# Change Log

## 0.14.2 - 2018-12-26

* Fix `db.getState is not a function`

## 0.14.1 - 2018-12-25

* Show error message if port is already used
* Upgrade to [lowdb](https://github.com/sumn2u/lowdb) `1.0`

## 0.14.0 - 2018-06-09

* Listen to `localhost` by default, instead of `0.0.0.0`

## 0.13.0 - 2018-05-30

* Bundle all index page assets so that you access it without network connection
* Drop Node 4 support

## 0.12.2 - 2018-04-26

* Add `_delay` query parameter
* Upgrade `please-upgrade-node` dependency

## 0.12.1 - 2017-11-02

* Disable logging for static content requests
* Remove bad `console.log`
* Update `dependencies`
* Use [`nanoid`](https://github.com/ai/nanoid)

## 0.12.0 - 2017-08-02

Re-include `body-parser` in `jsonServer.defaults()`

If you're using HAI Server in an Express server and experience issues, you can disable it by passing 

```js
jsonServer.defaults({ bodyParser: false })
```

## 0.11.2 - 2017-07-10

Fix `engines` field in `package.json`

## 0.11.1 - 2017-07-10

Add [please-upgrade-node](https://github.com/sumn2u/please-upgrade-node)

## 0.11.0 - 2017-07-05

Switch to [express-urlrewrite](https://github.com/kapouer/express-urlrewrite) to support rewriting query parameters (e.g.  `/articles?id=1 # → /posts/1`)

If you're rewriting default routes, you'll need to update your `routes.json` file
(see [add custom routes](https://github.com/sumn2u/hai-server#add-custom-routes) for updated doc).

## 0.10.3 - 2017-06-28

* Fix `line-break` error in CLI

## 0.10.2 - 2017-06-28

* Add `--foreignKeySuffix` option (e.g. snake case `post_id`) [#556](https://github.com/sumn2u/hai-server/pull/556) [#570](https://github.com/sumn2u/hai-server/pull/570)

## 0.10.1 - 2017-05-16

* Multiple fields sorting `GET /posts?_sort=user,views&_order=desc,asc`

## 0.10.0 - 2017-04-26

* __Drop Node `v0.12` support__
* Prevent `TypeError` when a filter is applied on a `null` value [#510](https://github.com/sumn2u/hai-server/issues/510)

## 0.9.6 - 2017-03-08

* Update index page
* Improve performances ([lowdb](https://github.com/sumn2u/lowdb) `v0.15`)
* Add `Location` header to newly created resources [#473](https://github.com/sumn2u/hai-server/pull/473)

## 0.9.5 - 2017-02-11

* Display custom routes on homepage
* Fix duplicate query params error [#352](https://github.com/sumn2u/hai-server/issues/352)

## 0.9.4 - 2016-12-08

* Improve rewriter [#431](https://github.com/sumn2u/hai-server/issues/431)
* Improve watch mode [#427](https://github.com/sumn2u/hai-server/pull/427)

## 0.9.3 - 2016-12-07

* Fix [#396](https://github.com/sumn2u/hai-server/issues/396) PUT/PATCH saves the updated item with an id that has been converted to string

## 0.9.2 - 2016-11-29

* Fix [#221](https://github.com/sumn2u/hai-server/issues/221) `nohup` support
* Fix [#420](https://github.com/sumn2u/hai-server/issues/420) TypeError when watching `db.json`

## 0.9.1 - 2016-11-21

* Fix
 * [#412](https://github.com/sumn2u/hai-server/issues/412)
 * [#451](https://github.com/sumn2u/hai-server/issues/411)

## 0.9.0 - 2016-11-11

* Shorter `uuid`
* No automatic conversion of strings to boolean or integer
* Create a default `db.json` file if it doesn't exist
* Fix
  * [#361](https://github.com/sumn2u/hai-server/issues/361)
  * [#363](https://github.com/sumn2u/hai-server/issues/363) [#365](https://github.com/sumn2u/hai-server/issues/365)
  * [#374](https://github.com/sumn2u/hai-server/issues/374)
  * [#383](https://github.com/sumn2u/hai-server/issues/383)
* Updated dependencies and codebase to ES6

## 0.8.23 - 2016-11-03

* Fix `Links` header

## 0.8.22 - 2016-10-04

* Fix `Links` header issue when using `_page`
* Add query params support to the route rewriter

## 0.8.21 - 2016-09-13

* Fix bodyParser issue when using custom routes

## 0.8.20 - 2016-09-12

* Fix [#355](https://github.com/sumn2u/hai-server/issues/355)
* Add `_page` support

## 0.8.19 - 2016-08-18

* Fix [#341](https://github.com/sumn2u/hai-server/issues/341)

## 0.8.18 - 2016-08-17

* Add CLI option `--middlewares` and support them in `hai-server.json` config file

## 0.8.17 - 2016-07-25

* Fix snapshot creation for JS files (ex: `hai-server generator.js`)

## 0.8.16 - 2016-07-11

* Support `x-www-form-urlencoded`

## 0.8.15 - 2016-07-03

* Bug fix: `--watch` option on OS X

## 0.8.14 - 2016-05-15

* Bug fix: data wasn't written to file in `v0.8.13` and `v0.8.12`

## 0.8.13 - 2016-05-12

* Make `_like` operator case insensitive

## 0.8.12 - 2016-05-08

* Minor bug fix

## 0.8.11 - 2016-05-08

* Support sort by nested field (e.g. `_sort=author.name`)
* Fix `graceful-fs` warning

## 0.8.10 - 2016-04-18

* CLI option `-ng/--no-gzip` to disable `gzip` compression

## 0.8.9 - 2016-03-17

* CLI can now read options from `hai-server.json` if present
* CLI option `-c/--config` to point to a different configuration file

## 0.8.8 - 2016-02-13

### Fixed

* Fix #233

## 0.8.7 - 2016-01-22

### Added

* `gzip` compression to improve performances
* CLI option `-nc/--no-cors` to disable CORS

## 0.8.6 - 2016-01-07

### Added

* CLI option `-ro/--read-only` to allow only GET requests

## 0.8.5 - 2015-12-28

### Fixed

* Fix #177

## 0.8.4 - 2015-12-13

### Added

* Like operator `GET /posts?title_like=json` (accepts RegExp)

## 0.8.3 - 2015-11-25

### Added

* CLI option `-q/--quiet`
* Nested route `POST /posts/1/comments`
* Not equal operator `GET /posts?id_ne=1`

## 0.8.2 - 2015-10-15

### Added

* CLI option `-S/--snapshots` to set a custom snapshots directory.

### Fixed

* Fix plural resources: `DELETE` should return `404` if resource doesn't exist.

## 0.8.1 - 2015-10-06

### Fixed

* Fix plural resources: `PUT` should replace resource instead of updating properties.
* Fix singular resources: `POST`, `PUT`, `PATCH` should not convert resource properties.

## 0.8.0 - 2015-09-21

### Changed

* `jsonServer.defaults` is now a function and can take an object.
If you're using the project as a module, you need to update your code:

```js
// Before
jsonServer.defaults

__tests__/cli/index.js
const fs = require('fs')
const path = require('path')
const cp = require('child_process')
const assert = require('assert')
const supertest = require('supertest')
const osTmpdir = require('os-tmpdir')
const tempWrite = require('temp-write')
const mkdirp = require('mkdirp')
const rimraf = require('rimraf')
const serverReady = require('server-ready')

let PORT = 3100

const middlewareFiles = {
  en: './../../__fixtures__/middlewares/en.js',
  jp: './../../__fixtures__/middlewares/jp.js',
  postbody: './../../__fixtures__/middlewares/postbody.js'
}

const bin = path.join(__dirname, '../../lib/cli/bin')

function cli(args) {
  return cp.spawn('node', ['--', bin, '-p', PORT].concat(args), {
    cwd: __dirname,
    stdio: ['pipe', process.stdout, process.stderr]
  })
}

describe('cli', () => {
  let child
  let request
  let dbFile
  let routesFile

  beforeEach(() => {
    dbFile = tempWrite.sync(
      JSON.stringify({
        posts: [{ id: 1 }, { _id: 2 }],
        comments: [{ id: 1, post_id: 1 }]
      }),
      'db.json'
    )

    routesFile = tempWrite.sync(
      JSON.stringify({ '/blog/*': '/$1' }),
      'routes.json'
    )

    ++PORT
    request = supertest(`http://localhost:${PORT}`)
  })

  afterEach(() => {
    child.kill('SIGKILL')
  })

  describe('db.json', () => {
    beforeEach(done => {
      child = cli([dbFile])
      serverReady(PORT, done)
    })

    test('should support JSON file', done => {
      request.get('/posts').expect(200, done)
    })

    test('should send CORS headers', done => {
      const origin = 'http://example.com'

      request
        .get('/posts')
        .set('Origin', origin)
        .expect('access-control-allow-origin', origin)
        .expect(200, done)
    })

    test('should update JSON file', done => {
      request
        .post('/posts')
        .send({ title: 'hello' })
        .end(() => {
          setTimeout(() => {
            const str = fs.readFileSync(dbFile, 'utf8')
            assert(str.indexOf('hello') !== -1)
            done()
          }, 1000)
        })
    })
  })

  describe('seed.js', () => {
    beforeEach(done => {
      child = cli(['../../__fixtures__/seed.js'])
      serverReady(PORT, done)
    })

    test('should support JS file', done => {
      request.get('/posts').expect(200, done)
    })
  })

  describe('remote db', () => {
    beforeEach(done => {
      child = cli(['https://jsonplaceholder.sumn2u.com/db'])
      serverReady(PORT, done)
    })

    test('should support URL file', done => {
      request.get('/posts').expect(200, done)
    })
  })

  describe('db.json -r routes.json -m middleware.js -i _id --foreignKeySuffix _id --read-only', () => {
    beforeEach(done => {
      child = cli([
        dbFile,
        '-r',
        routesFile,
        '-m',
        middlewareFiles.en,
        '-i',
        '_id',
        '--read-only',
        '--foreignKeySuffix',
        '_id'
      ])
      serverReady(PORT, done)
    })

    test('should use routes.json and _id as the identifier', done => {
      request.get('/blog/posts/2').expect(200, done)
    })

    test('should use _id as foreignKeySuffix', async () => {
      const response = await request.get('/posts/1/comments')
      assert.strictEqual(response.body.length, 1)
    })

    test('should apply middlewares', done => {
      request.get('/blog/posts/2').expect('X-Hello', 'World', done)
    })

    test('should allow only GET requests', done => {
      request.post('/blog/posts').expect(403, done)
    })
  })

  describe('db.json -m first-middleware.js second-middleware.js', () => {
    beforeEach(done => {
      child = cli([dbFile, '-m', middlewareFiles.en, middlewareFiles.jp])
      serverReady(PORT, done)
    })

    test('should apply all middlewares', done => {
      request
        .get('/posts')
        .expect('X-Hello', 'World')
        .expect('X-Konnichiwa', 'Sekai', done)
    })
  })

  describe('db.json -m postbody-middleware.js', () => {
    beforeEach(done => {
      child = cli([dbFile, '-m', middlewareFiles.postbody])
      serverReady(PORT, done)
    })

    test('should have post body in middleware', done => {
      request
        .post('/posts')
        .send({ name: 'test' })
        .expect('name', 'test', done)
    })
  })

  describe('db.json -d 1000', () => {
    beforeEach(done => {
      child = cli([dbFile, '-d', 1000])
      serverReady(PORT, done)
    })

    test('should delay response', done => {
      const start = new Date()
      request.get('/posts').expect(200, function(err) {
        const end = new Date()
        done(end - start > 1000 ? err : new Error("Request wasn't delayed"))
      })
    })
  })

  describe('db.json -s ../../__fixtures__/public -S /some/path/snapshots', () => {
    const snapshotsDir = path.join(osTmpdir(), 'snapshots')
    const publicDir = '../../__fixtures__/public'

    beforeEach(done => {
      rimraf.sync(snapshotsDir)
      mkdirp.sync(snapshotsDir)

      child = cli([dbFile, '-s', publicDir, '-S', snapshotsDir])
      serverReady(PORT, () => {
        child.stdin.write('s\n')
        setTimeout(done, 100)
      })
    })

    test('should serve ../../__fixtures__/public', done => {
      request.get('/').expect(/Hello/, done)
    })

    test('should save a snapshot in snapshots dir', () => {
      assert.strictEqual(fs.readdirSync(snapshotsDir).length, 1)
    })
  })

  describe('../../__fixtures__/seed.json --no-cors=true', () => {
    beforeEach(done => {
      child = cli(['../../__fixtures__/seed.js', '--no-cors=true'])
      serverReady(PORT, done)
    })

    test('should not send Access-Control-Allow-Origin headers', done => {
      const origin = 'http://example.com'

      request
        .get('/posts')
        .set('Origin', origin)
        .expect(200)
        .end((err, res) => {
          if (err) {
            done(err)
          }
          if ('access-control-allow-origin' in res.headers) {
            done(new Error('CORS headers were not excluded from response'))
          } else {
            done()
          }
        })
    })
  })

  describe('../../__fixtures__/seed.json --no-gzip=true', () => {
    beforeEach(done => {
      child = cli(['../../__fixtures__/seed.js', '--no-gzip=true'])
      serverReady(PORT, done)
    })

    test('should not set Content-Encoding to gzip', done => {
      request
        .get('/posts')
        .expect(200)
        .end(function(err, res) {
          if (err) {
            done(err)
          } else if ('content-encoding' in res.headers) {
            done(new Error('Content-Encoding is set to gzip'))
          } else {
            done()
          }
        })
    })
  })

  describe('--watch db.json -r routes.json', () => {
    beforeEach(done => {
      child = cli([dbFile, '-r', routesFile, '--watch'])
      serverReady(PORT, done)
    })

    test('should watch db file', done => {
      fs.writeFileSync(dbFile, JSON.stringify({ foo: [] }))
      setTimeout(() => {
        request.get('/foo').expect(200, done)
      }, 1000)
    })

    test('should watch routes file', done => {
      fs.writeFileSync(routesFile, JSON.stringify({ '/api/*': '/$1' }))
      setTimeout(() => {
        request.get('/api/posts').expect(200, done)
      }, 1000)
    })
  })

  describe('non existent db.json', () => {
    beforeEach(done => {
      fs.unlinkSync(dbFile)
      child = cli([dbFile])
      serverReady(PORT, done)
    })

    test("should create JSON file if it doesn't exist", done => {
      request.get('/posts').expect(200, done)
    })
  })

  describe('db.json with error', () => {
    beforeEach(() => {
      dbFile = tempWrite.sync(JSON.stringify({ 'a/b': [] }), 'db-error.json')
    })

    test('should exit with an error', done => {
      child = cli([dbFile])
      child.on('exit', code => {
        if (code === 1) {
          return done()
        }
        return done(new Error('should exit with error code'))
      })
    })
  })
})


src/cli/run.js
const fs = require('fs')
const path = require('path')
const jph = require('json-parse-helpfulerror')
const _ = require('lodash')
const chalk = require('chalk')
const enableDestroy = require('server-destroy')
const pause = require('connect-pause')
const is = require('./utils/is')
const load = require('./utils/load')

const haijsServer = require('../server').haijsServer


function prettyPrint(argv, object, rules) {
  const root = `http://${argv.host}:${argv.port}`

  console.log()
  console.log(chalk.bold('  Resources'))
  for (let prop in object) {
    console.log(`  ${root}/${prop}`)
  }

  if (rules) {
    console.log()
    console.log(chalk.bold('  Other routes'))
    for (var rule in rules) {
      console.log(`  ${rule} -> ${rules[rule]}`)
    }
  }

  console.log()
  console.log(chalk.bold('  Home'))
  console.log(`  ${root}`)
  console.log()
}

function createApp(db, routes, middlewares, auth, argv) {
  const app = haijsServer.create()

  const { foreignKeySuffix } = argv

  let router = haijsServer.router(
    db,
    foreignKeySuffix ? { foreignKeySuffix } : undefined,
    auth
  )

  const defaultsOpts = {
    logger: !argv.quiet,
    readOnly: argv.readOnly,
    noCors: argv.noCors,
    noGzip: argv.noGzip,
    bodyParser: true
  }

  if (argv.static) {
    defaultsOpts.static = path.join(process.cwd(), argv.static)
  }

  const defaults = haijsServer.defaults(defaultsOpts)
  app.use(defaults)

  if (routes) {
    const rewriter = haijsServer.rewriter(routes)
    app.use(rewriter)
  }

  if (middlewares) {
    app.use(middlewares)
  }

  if (argv.delay) {
    app.use(pause(argv.delay))
  }
  if (auth) {
    let authenticationURL = auth.authenticationURL || '/auth/login'
    let authenticatedUsers = auth.users || []
    // Use this for access token
    router.post(authenticationURL, (req, res) => {
      const { email, password } = req.body
      if (isAuthenticated({ email, password }, authenticatedUsers) === false) {
        const status = 401
        const message = 'Incorrect email or password'
        res.status(status).json({ status, message })
        return
      }
      if (isAuthenticated({ email, password }, authenticatedUsers ) === false) {
        const access_token = createToken({ email, password })
        res.status(200).json({ access_token })
       }
    })
  }

  router.db._.id = argv.id
  app.db = router.db
  app.use(router)

  return app
}

module.exports = function(argv) {
  const source = argv._[0]
  let app
  let server

  if (!fs.existsSync(argv.snapshots)) {
    console.log(`Error: snapshots directory ${argv.snapshots} doesn't exist`)
    process.exit(1)
  }

  // noop log fn
  if (argv.quiet) {
    console.log = () => {}
  }

  console.log()
  console.log(chalk.cyan('  \\{^_^}/ hi!'))

  function start(cb) {
    console.log()

    console.log(chalk.gray('  Loading', source))

    server = undefined

    // create db and load object, JSON file, JS or HTTP database
    return load(source).then(db => {
      // Load additional routes
      let routes
      if (argv.routes) {
        console.log(chalk.gray('  Loading', argv.routes))
        routes = JSON.parse(fs.readFileSync(argv.routes))
      }

      // Load middlewares
      let middlewares
      if (argv.middlewares) {
        middlewares = argv.middlewares.map(function(m) {
          console.log(chalk.gray('  Loading', m))
          return require(path.resolve(m))
        })
      }
      // Load auth
      let auth
      if (argv.auth) {
        console.log(chalk.gray('Loading', argv.auth))
        auth = JSON.parse(fs.readFileSync(argv.auth))
      }
      // Done
      console.log(chalk.gray('  Done'))

      // Create app and server
      app = createApp(db, routes, middlewares, auth, argv)
      server = app.listen(argv.port, argv.host)

      // Enhance with a destroy function
      enableDestroy(server)

      // Display server informations
      prettyPrint(argv, db.getState(), routes)

      // Catch and handle any error occurring in the server process
      process.on('uncaughtException', error => {
        if (error.errno === 'EADDRINUSE')
          console.log(
            chalk.red(
              `Cannot bind to the port ${
                error.port
              }. Please specify another port number either through --port argument or through the hai-server.json configuration file`
            )
          )
        else console.log('Some error occurred', error)
        process.exit(1)
      })
    })
  }

  // Start server
  start()
    .then(() => {
      // Snapshot
      console.log(
        chalk.gray(
          '  Type s + enter at any time to create a snapshot of the database'
        )
      )

      // Support nohup
      // https://github.com/sumn2u/hai-server/issues/221
      process.stdin.on('error', () => {
        console.log(`  Error, can't read from stdin`)
        console.log(`  Creating a snapshot from the CLI won't be possible`)
      })
      process.stdin.setEncoding('utf8')
      process.stdin.on('data', chunk => {
        if (chunk.trim().toLowerCase() === 's') {
          const filename = `db-${Date.now()}.json`
          const file = path.join(argv.snapshots, filename)
          const state = app.db.getState()
          fs.writeFileSync(file, JSON.stringify(state, null, 2), 'utf-8')
          console.log(
            `  Saved snapshot to ${path.relative(process.cwd(), file)}\n`
          )
        }
      })

      // Watch files
      if (argv.watch) {
        console.log(chalk.gray('  Watching...'))
        console.log()
        const source = argv._[0]

        // Can't watch URL
        if (is.URL(source)) throw new Error("Can't watch URL")

        // Watch .js or .json file
        // Since lowdb uses atomic writing, directory is watched instead of file
        const watchedDir = path.dirname(source)
        let readError = false
        fs.watch(watchedDir, (event, file) => {
          // https://github.com/sumn2u/hai-server/issues/420
          // file can be null
          if (file) {
            const watchedFile = path.resolve(watchedDir, file)
            if (watchedFile === path.resolve(source)) {
              if (is.FILE(watchedFile)) {
                let obj
                try {
                  obj = jph.parse(fs.readFileSync(watchedFile))
                  if (readError) {
                    console.log(chalk.green(`  Read error has been fixed :)`))
                    readError = false
                  }
                } catch (e) {
                  readError = true
                  console.log(chalk.red(`  Error reading ${watchedFile}`))
                  console.error(e.message)
                  return
                }

                // Compare .json file content with in memory database
                const isDatabaseDifferent = !_.isEqual(obj, app.db.getState())
                if (isDatabaseDifferent) {
                  console.log(
                    chalk.gray(`  ${source} has changed, reloading...`)
                  )
                  server && server.destroy(() => start())
                }
              }
            }
          }
        })

        // Watch routes
        if (argv.routes) {
          const watchedDir = path.dirname(argv.routes)
          fs.watch(watchedDir, (event, file) => {
            if (file) {
              const watchedFile = path.resolve(watchedDir, file)
              if (watchedFile === path.resolve(argv.routes)) {
                console.log(
                  chalk.gray(`  ${argv.routes} has changed, reloading...`)
                )
                server && server.destroy(() => start())
              }
            }
          })
        }
      }
    })
    .catch(err => {
      console.log(err)
      process.exit(1)
    })
}
